src/main/java/org/opensearch/knn/search/NativeEngineKnnFloatVectorQuery.java

```diff
@@ -0,0 +1,112 @@
+            // If filter query is null and accepted doc are not null then we can directly convert it to bitset
+            // This maintains the status quo with current code path wherein deleted docs are not considered
+            // Without this check deleted docs will be passed in and impact latencies.
+            filterBitset = (BitSet) acceptDocs;
+            cardinality = filterBitset.cardinality();
+        }
+
+        Map<Integer, Float> leafDocScores = knnWeight.doANNSearch(context, filterBitset, cardinality);
+        if (leafDocScores == null) {
+            leafDocScores = Collections.emptyMap();
+        }
+        final Bits liveDocs = context.reader().getLiveDocs();
+
+        final List<Map.Entry<Integer, Float>> topScores = new ArrayList<>(leafDocScores.entrySet());
+        topScores.sort(Map.Entry.<Integer, Float>comparingByValue().reversed());
+
+        //This is to trick the implementation to force exactsearch after approxsearch
+        TotalHits.Relation relation = query.getFilterQuery() != null
+                && cardinality >= query.getK() && topScores.size() < query.getK()
+                ? TotalHits.Relation.GREATER_THAN_OR_EQUAL_TO
+                : TotalHits.Relation.EQUAL_TO;
+
+        return convertDocScoresToTopDocs(topScores, liveDocs, context.docBase, relation);
+    }
+
+    protected TopDocs exactSearch(LeafReaderContext context, DocIdSetIterator acceptIterator, QueryTimeout queryTimeout) {
+        if (acceptIterator instanceof BitSetIterator) {
+            BitSetIterator bitSetIterator = (BitSetIterator) acceptIterator;
+            BitSet acceptedBitSet = bitSetIterator.getBitSet();
+
+            final Map<Integer, Float> docToScore = knnWeight.doExactSearch(context, acceptedBitSet, acceptedBitSet.cardinality());
+            return convertDocScoresToTopDocs(new ArrayList<>(docToScore.entrySet()), null, context.docBase, TotalHits.Relation.EQUAL_TO);
+        }
+        throw new IllegalStateException("DocIdSetIterator is not a BitSetIterator");
+    }
+
+    private TopDocs convertDocScoresToTopDocs(final List<Map.Entry<Integer, Float>> docScores, final Bits liveDocs, int docBase, TotalHits.Relation relation) {
+        final List<ScoreDoc> scoreDocs = new ArrayList<>(docScores.size());
+        int totalHits = 0;
+        for (final Map.Entry<Integer, Float> entry : docScores) {
+            // since lucene query executes at shard, we need to filter delete docs
+            if (liveDocs == null || liveDocs.get(entry.getKey())) {
+                ScoreDoc scoreDoc = new ScoreDoc(entry.getKey() + docBase, entry.getValue());
+                scoreDocs.add(scoreDoc);
+                totalHits++;
+            }
+        }
+
+        return new TopDocs(new TotalHits(totalHits, relation), scoreDocs.toArray(ScoreDoc[]::new));
+    }
+
+    @Override
+    protected TopDocs mergeLeafResults(TopDocs[] perLeafResults) {
+        // Rescore here
```

Comment:
Issue with re-scoring here is we need access to the LeafReaderContext for each set of results to lookup the vectors. So, not sure we can re-score here. 

Instead, I think we will need to re-score from rewrite. If we want to share functionality, I think we could call super.rewrite(), get the resulting Query, and then rescore. Otherwise, we should just override the rewrite method.

Reply:
Ah I see there is not way to cleanly rescore in this approach then